File: examples/Csharp.ExampleApplication/Hooks/Transformers/MigratedTagTransformer.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Tableau.Migration;
using Tableau.Migration.Content;
using Tableau.Migration.Engine.Hooks.Transformers;
using Tableau.Migration.Resources;

namespace Csharp.ExampleApplication.Hooks.Transformers
{
    #region class
    public class MigratedTagTransformer<T> : ContentTransformerBase<T> where T : IContentReference, IWithTags
    {
        private readonly ILogger<IContentTransformer<T>>? _logger;

        public MigratedTagTransformer(ISharedResourcesLocalizer localizer, ILogger<IContentTransformer<T>> logger) : base(localizer, logger)
        {
            _logger = logger;
        }

        public override async Task<T?> TransformAsync(T itemToTransform, CancellationToken cancel)
        {
            var tag = "Migrated";

            // Add the tag to the content item.
            itemToTransform.Tags.Add(new Tag(tag));

            _logger?.LogInformation(
                @"Added ""{Tag}"" tag to {ContentType} {ContentLocation}.",
                tag,
                typeof(T).Name,
                itemToTransform.Location);

            return await Task.FromResult(itemToTransform);
        }

        public async Task<IPublishableWorkbook?> TransformAsync(IPublishableWorkbook ctx, CancellationToken cancel)
            => await TransformAsync(ctx, cancel);

        public async Task<IPublishableDataSource?> TransformAsync(IPublishableDataSource ctx, CancellationToken cancel)
            => await TransformAsync(ctx, cancel);
    }
    #endregion
}

File: examples/Python.ExampleApplication/Hooks/transformers/migrated_tag_transformer.py
from typing import TypeVar
from tableau_migration import (
    ContentTransformerBase,
    IPublishableDataSource,
    IPublishableWorkbook)
from tableau_migration.migration_content import PyTag

T = TypeVar("T")


class MigratedTagTransformer(ContentTransformerBase[T]):
    def transform(self, itemToTransform: T) -> T:
        tag: str = "Migrated"
        
        new_tag = PyTag.create(tag)
        
        current_tags = list(itemToTransform.tags)
        current_tags.append(new_tag)
        
        itemToTransform.tags = current_tags
        
        return itemToTransform
    
class MigratedTagTransformerForDataSources(MigratedTagTransformer[IPublishableDataSource]):
    pass

class MigratedTagTransformerForWorkbooks(MigratedTagTransformer[IPublishableWorkbook]):
    pass